### ISSUE TITLE
t.i.tcp should not use os.strerror for Windows socket error codes

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2425]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)| @ghazel reported|
|-|-|
|Trac ID|trac#2425|
|Type|defect|
|Created|2007-01-29 10:13:58Z|

os.strerror() is not appropriate for Windows socket error codes.

os.strerror(10061) #=> 'Unknown error'

There is a different native function (FormatMessage) to get the real error message. However any method of calling that would introduce a dependency (win32api, ctypes, take your pick). In the interest of avoiding that, several of the common WSA error codes are in the socket module in socket.errorTab. It's not documented, but it's been around for almost 6 years:
http://sourceforge.net/tracker/?func=detail&atid=105470&aid=406642&group_id=5470
And hey look, moshez.

So something like this would work better:
```
def strerror(err):
    errorTab = getattr(socket, "errorTab", {})
    if err in errorTab:
        return errorTab[err]
    return os.strerror(err)
```

strerror(10061) #=> 'Connection refused.'

The question is, where should it go? It almost seems like a t.python thing, but only t.i.tcp uses it in this way.

Attachments:

* [formatError.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a91/a91c1dfe80539381caf15228e9b6d163f3268a7c/8e1e816db224237f3286fa7c7da6822d05c65227.py) (1020 bytes) - added by arkanes on 2007-03-17 03:51:12Z - proper error code lookup on Windows
* [formatError.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a91/a91c1dfe80539381caf15228e9b6d163f3268a7c/89efc890747918c2875164bdb728602fbfd0aae6.patch) (5542 bytes) - added by arkanes on 2007-03-18 12:49:33Z - patch adds formatError, unit tests
* [formatError2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a91/a91c1dfe80539381caf15228e9b6d163f3268a7c/0408746dcccc4c08bf12d3160b55a1b9c9b5d4c6.patch) (5834 bytes) - added by ghazel on 2007-10-28 01:36:05Z - formatError patch, take 2

<details><summary>Searchable metadata</summary>

```
trac-id__2425 2425
type__defect defect
reporter__ghazel ghazel
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1170065638000000 1170065638000000
changetime__1194617048000000 1194617048000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="arkanes's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|arkanes commented|
|-|-|

The *real* fix is that os.strerror needs to use FormatMessage on Windows. In the meantime, I'm attaching a function that can be used to look up error codes through with fallback through win32, ctypes, the errorTab in socket module, and finally os.strerror.

I'm not sure what kind of unit tests would be appropriate for this - it's hard to test conditional module imports.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph set owner to arkanes|
|-|-|

I'm not the windows maintainer, so I can't really do a full review here, but I can tell you that the fact that there are no unit tests will be an instant reject when someone does do a thorough review, so you should probably add some.

**Comment by twisted-trac** [comment]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)<a name="note_4"></a>|@ghazel commented|
|-|-|

This is much nicer than the original suggestion. Thanks arkanes.

**Comment by twisted-trac** [comment]
|<img alt="arkanes's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|arkanes commented|
|-|-|

Added a couple unit tests to make sure the lookup works and that it doesn't bomb out on other platforms.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

Rather than having a conditional inside the test case, why not have two separate test cases, one for win32 platforms with the fallback and one without?  Skip them as appropriate.

Sorry to bug you about trivia, but:

  * Each test must have a docstring, explaining what it is that it tests.
  * The convention for test case names is actually ```test_fooBarBaz```, not ```testFooBarBaz```.  Unfortunately a large number of tests in Twisted haven't had their names cleaned up yet, but that is the official standard.

**Comment by twisted-trac** [comment]
|<img alt="arkanes's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|arkanes commented|
|-|-|

Replying to [glyph](#note_6):
> Rather than having a conditional inside the test case, why not have two separate test cases, one for win32 platforms with the fallback and one without?  Skip them as appropriate.
>
> Sorry to bug you about trivia, but:
>
>   * Each test must have a docstring, explaining what it is that it tests.
>   * The convention for test case names is actually ```test_fooBarBaz```, not ```testFooBarBaz```.  Unfortunately a large number of tests in Twisted haven't had their names cleaned up yet, but that is the official standard.


Patch updated - comments in tests moved to docstrings, docstrings on all tests, naming convention updated. Changed the first test to two tests with individual skips as you suggested.

**Comment by twisted-trac** [comment]
|<img alt="arkanes's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|arkanes commented|
|-|-|

Misread naming convention, patch updated.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set owner to arkanes|
|-|-|

Quick review:
 * Please use ```try: import; except ImportError:...; else: action``` instead of ```try: import; action; except ImportError...```
 * the skip use the wrong case of methods (test_FormatMessageSafeToCall.skip instead of test_formatMessageSafeToCall.skip)
 * Update copyright of the 3 files touched.
 * Use ```from twisted.python import win32``` instead of ```import twisted.python.win32 as win32```
 * The tests are a bit disturbing... The first is run on non-win32, whereas the test file is test_win32. The second one is not really useful. And the third one makes actually the same calls as the original function. For example if formatError returns None, your tests will pass. I don't have a answer though, so it could probably go like this.

**Comment by twisted-trac** [comment]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)<a name="note_12"></a>|@ghazel set owner to @therve|
|-|-|

formatError2.patch addresses the review issues mentioned by therve. It's been 7 months, so I figured I'd take a crack at getting this finished.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve commented|
|-|-|

(In [21460]) Apply tracker's patch with some fixes.

Refs [#2425](https://github.com/twisted/twisted/issues/2425)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve removed owner|
|-|-|

I got my hands dirty by creating the branch win32-strerror-2425.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

formatError needs more tests.  I'm not exactly sure what they'll look like yet.  I'll try to write them.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Okay I did something in that branch.  I have no Windows environment to test with, and all the Windows slaves are offline right now, so I'm done with this for now.  Maybe it works, maybe not, someone else can figure it out.

**Comment by twisted-trac** [comment]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)<a name="note_17"></a>|@ghazel commented|
|-|-|

Tests pass after exarkun's changes, and the function seems to still work too.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

thanks for checking this over ghazel.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_19"></a>|@therve set owner to @exarkun|
|-|-|

OK, I merged forward to **win32-strerror-2425-2**, buildbot is green, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun set status to `closed`|
|-|-|

(In [21682]) Merge win32-strerror-2425-2

Author: arkanes, ghazel, exarkun
Reviewer: therve
Fixes [#2425](https://github.com/twisted/twisted/issues/2425)

Change twisted.internet.tcp to use Windows-aware error code lookup when creating
exceptions for connection failure.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
